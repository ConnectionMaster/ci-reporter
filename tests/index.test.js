const {createRobot} = require('probot')
const nock = require('nock')
const path = require('path')
const fs = require('fs')
const ciReporter = require('../functions/src')

const readFile = file => fs.readFileSync(path.join(__dirname, 'fixtures', file), 'utf8')
const commentsGet = require('./fixtures/issues.getComments.json')
const commentsGetTwo = require('./fixtures/issues.getComments-two.json')

describe('ci-reporter', () => {
  let robot, github

  beforeEach(() => {
    robot = createRobot()
    github = {
      issues: {
        getComments: jest.fn(() => Promise.resolve({data: []})),
        createComment: jest.fn(),
        editComment: jest.fn()
      },
      repos: {
        getContent: jest.fn(() => Promise.resolve({data: { content: '' }}))
      }
    }

    robot.auth = jest.fn(() => Promise.resolve(github))
    ciReporter(robot)
  })

  describe('Travis CI', () => {
    it('creates the correct comment', async () => {
      const log = readFile(path.join('travis', 'log.txt'))

      nock('https://api.travis-ci.org')
        .get('/build/123').reply(200, {
          pull_request_number: 1,
          jobs: [{ id: 1234, number: 1, state: 'failed' }]
        })
        .get('/job/1234/log').reply(200, { content: log })

      const event = {
        event: 'status',
        payload: {
          sha: 'b04b9ce383a933ed1a0a7b3de9e1cd31770b380e',
          target_url: 'https://travis-ci.org/JasonEtco/public-test/builds/123?utm_source=github_status&utm_medium=notification',
          context: 'continuous-integration/travis-ci/pr',
          state: 'failure',
          repository: {
            name: 'public-test',
            owner: { login: 'JasonEtco' }
          },
          installation: { id: 123 }
        }
      }

      await robot.receive(event)
      expect(github.issues.createComment).toHaveBeenCalledTimes(1)

      const args = github.issues.createComment.mock.calls[0]

      expect(args[0].body).toMatchSnapshot()
      expect(args[0].number).toBe(1)
      expect(args[0].owner).toBe('JasonEtco')
      expect(args[0].repo).toBe('public-test')
    })
  })

  describe('Circle CI', () => {
    let event, build, output, commit

    beforeEach(() => {
      event = {
        event: 'status',
        payload: {
          sha: 'b04b9ce383a933ed1a0a7b3de9e1cd31770b380e',
          target_url: 'https://circleci.com/gh/JasonEtco/todo/5?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link',
          context: 'ci/circleci',
          state: 'failure',
          repository: {
            name: 'todo',
            owner: { login: 'JasonEtco' }
          },
          installation: { id: 123 }
        }
      }

      build = readFile(path.join('circle', 'build.json'))
      output = readFile(path.join('circle', 'output.json'))
      commit = readFile(path.join('circle', 'commit.json'))

      nock('https://circleci.com')
        .get('/api/v1.1/project/github/JasonEtco/todo/5').reply(200, build)
        .get('/fake-output-url').reply(200, output)
    })

    it('creates the correct comment', async () => {
      await robot.receive(event)
      expect(github.issues.createComment).toHaveBeenCalledTimes(1)

      const args = github.issues.createComment.mock.calls[0]

      expect(args[0].body).toMatchSnapshot()
      expect(args[0].number).toBe(1)
      expect(args[0].owner).toBe('JasonEtco')
      expect(args[0].repo).toBe('todo')
    })

    it('creates the correct comment with before/after disabled', async () => {
      github.repos.getContent.mockReturnValueOnce(Promise.resolve({ data: {content: Buffer.from('after: false\nbefore: false')} }))
      await robot.receive(event)
      expect(github.issues.createComment.mock.calls[0][0].body).toMatchSnapshot()
    })

    it('creates the correct comment with custom before/after', async () => {
      github.repos.getContent.mockReturnValueOnce(Promise.resolve({ data: {content: Buffer.from('before: I come before!\nafter: And I come after!')} }))
      await robot.receive(event)
      expect(github.issues.createComment.mock.calls[0][0].body).toMatchSnapshot()
    })

    it('does not create a comment if the status is not in a PR', async () => {
      nock.cleanAll()
      nock('https://circleci.com')
        .get('/api/v1.1/project/github/JasonEtco/todo/5').reply(200, commit)
        .get('/fake-output-url').reply(200, output)

      await robot.receive(event)
      expect(github.issues.createComment).not.toHaveBeenCalled()
    })

    it('updates an existing comment', async () => {
      github.issues.getComments.mockReturnValueOnce(Promise.resolve(commentsGet))
      await robot.receive(event)
      expect(github.issues.editComment.mock.calls[0][0]).toMatchSnapshot()
    })

    it('updates an existing comment twice', async () => {
      nock.cleanAll()
      nock('https://circleci.com')
        .get('/api/v1.1/project/github/JasonEtco/todo/5').times(2).reply(200, build)
        .get('/fake-output-url').times(2).reply(200, output)

      github.issues.getComments.mockReturnValueOnce(Promise.resolve(commentsGet))
      await robot.receive(event)
      github.issues.getComments.mockReturnValueOnce(Promise.resolve(commentsGetTwo))
      await robot.receive(event)
      expect(github.issues.editComment.mock.calls[1][0]).toMatchSnapshot()
    })

    it('respect the updateComment config', async () => {
      github.repos.getContent.mockReturnValueOnce(Promise.resolve({data: {content: Buffer.from('updateComment: false')}}))
      github.issues.getComments.mockReturnValueOnce(Promise.resolve(commentsGet))
      await robot.receive(event)
      expect(github.issues.createComment).toHaveBeenCalled()
      expect(github.issues.editComment).not.toHaveBeenCalled()
    })
  })

  it('does nothing if the status context is not accounted for', async () => {
    const event = {
      event: 'status',
      payload: {
        context: 'i/do/not/exist',
        state: 'failure',
        installation: { id: 123 },
        repository: {
          name: 'public-test',
          owner: { login: 'JasonEtco' }
        }
      }
    }

    await robot.receive(event)
    expect(github.issues.createComment).not.toHaveBeenCalled()
  })

  it('does nothing on non-failed status', async () => {
    const event = {
      event: 'status',
      payload: {
        state: 'passed',
        installation: { id: 123 },
        repository: {
          name: 'public-test',
          owner: { login: 'JasonEtco' }
        }
      }
    }

    await robot.receive(event)
    expect(github.issues.createComment).not.toHaveBeenCalled()
  })
})
