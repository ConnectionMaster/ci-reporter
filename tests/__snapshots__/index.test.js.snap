// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ci-reporter Circle CI creates the correct comment 1`] = `
"### The build is failing

âœ¨ Good work on this PR so far! âœ¨ Unfortunately, the [Circle CI build](https://circleci.com/gh/JasonEtco/todo/5?utm_campaign&#x3D;vcs-integration-link&amp;utm_medium&#x3D;referral&amp;utm_source&#x3D;github-build-link) is failing. Here's the output:

##### \`npm test\`

\`\`\`
> todo@1.0.0 test /home/circleci/repo
> jest && codecov && standard

[0m[7m[1m[32m PASS [39m[22m[27m[0m [2m__tests__/lib/[22m[1mmetadata.js[22m
[0m[7m[1m[32m PASS [39m[22m[27m[0m [2m__tests__/lib/[22m[1mgenerate-blob-link.js[22m
[0m[7m[1m[32m PASS [39m[22m[27m[0m [2m__tests__/lib/[22m[1mreopen-closed.js[22m
[0m[7m[1m[32m PASS [39m[22m[27m[0m [2m__tests__/lib/[22m[1mcommit-is-in-pr.js[22m
[0m[7m[1m[32m PASS [39m[22m[27m[0m [2m__tests__/lib/[22m[1mgenerate-label.js[22m
[0m[7m[1m[32m PASS [39m[22m[27m[0m [2m__tests__/lib/[22m[1mgenerate-body.js[22m
[0m[7m[1m[32m PASS [39m[22m[27m[0m [2m__tests__/[22m[1mlanguages.js[22m
[0m[7m[1m[32m PASS [39m[22m[27m[0m [2m__tests__/[22m[1minstallation.js[22m
[0m[7m[1m[32m PASS [39m[22m[27m[0m [2m__tests__/[22m[1mpr-comments.js[22m
[0m[7m[1m[32m PASS [39m[22m[27m[0m [2m__tests__/[22m[1mmerge-handler.js[22m
[0m[7m[1m[32m PASS [39m[22m[27m[0m [2m__tests__/[22m[1mopen-issues.js[22m
[999D[K
[1mTest Suites: [22m[1m[32m11 passed[39m[22m, 11 total
[1mTests:       [22m[1m[32m70 passed[39m[22m, 70 total
[1mSnapshots:   [22m0 total
[1mTime:[22m        4.335s
[2mRan all test suites[22m[2m.[22m
--------------------------------|----------|----------|----------|----------|----------------|
File                            |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
--------------------------------|----------|----------|----------|----------|----------------|
All files                       |      100 |      100 |      100 |      100 |                |
 repo                           |      100 |      100 |      100 |      100 |                |
  index.js                      |      100 |      100 |      100 |      100 |                |
 repo/lib                       |      100 |      100 |      100 |      100 |                |
  check-for-duplicate-issue.js  |      100 |      100 |      100 |      100 |                |
  commit-is-in-pr.js            |      100 |      100 |      100 |      100 |                |
  default-config.js             |      100 |      100 |      100 |      100 |                |
  generate-blob-link.js         |      100 |      100 |      100 |      100 |                |
  generate-body.js              |      100 |      100 |      100 |      100 |                |
  generate-label.js             |      100 |      100 |      100 |      100 |                |
  get-file-contents.js          |      100 |      100 |      100 |      100 |                |
  helpers.js                    |      100 |      100 |      100 |      100 |                |
  merge-handler.js              |      100 |      100 |      100 |      100 |                |
  metadata.js                   |      100 |      100 |      100 |      100 |                |
  open-issues.js                |      100 |      100 |      100 |      100 |                |
  organize-commits.js           |      100 |      100 |      100 |      100 |                |
  reopen-closed.js              |      100 |      100 |      100 |      100 |                |
  search-contents-for-titles.js |      100 |      100 |      100 |      100 |                |
--------------------------------|----------|----------|----------|----------|----------------|
  _____          _  
 / ____|        | |  
| |     ___   __| | ___  ___ _____   __  
| |    / _ \\\\ / _\` |/ _ \\\\/ __/ _ \\\\ \\\\ / /  
| |___| (_) | (_| |  __/ (_| (_) \\\\ V /  
 \\\\_____\\\\___/ \\\\__,_|\\\\___|\\\\___\\\\___/ \\\\_/  
                                v2.3.0
==> Detecting CI Provider
    Circle CI Detected
==> Configuration: 
    Endpoint: https://codecov.io
{ service: 'circleci',
  build: '4.0',
  job: '4.0',
  commit: '36eeffbb45a75f1ac2ef3c8b491b5a1d9b4afefa',
  branch: 'cleanup',
  pr: undefined,
  slug: 'JasonEtco/todo',
  package: 'node-v2.3.0' }
==> Building file structure
==> Generating gcov reports (skip via --disable=gcov)
    $ find /home/circleci/repo -type f -name '*.gcno'  -exec gcov  {} +
==> Scanning for reports
    + /home/circleci/repo/coverage/clover.xml
    + /home/circleci/repo/coverage/lcov.info
==> Uploading reports
    Success!
    View report at: https://codecov.io/github/JasonEtco/todo/commit/36eeffbb45a75f1ac2ef3c8b491b5a1d9b4afefa
standard: Use JavaScript Standard Style (https://standardjs.com)
standard: Run \`standard --fix\` to automatically fix some problems.
  /home/circleci/repo/lib/search-contents-for-titles.js:15:2: Newline required at end of file but not found.
\`\`\`

I'm sure you can fix it! If you need help, don't hesitate to ask a maintainer of the project!

---

###### This comment was automagically generated by [ci-reporter](https://github.com/apps/ci-reporter). If you see a problem, open [an issue here](https://github.com/jasonetco/ci-reporter)."
`;

exports[`ci-reporter Travis CI creates the correct comment 1`] = `
"### The build is failing

âœ¨ Good work on this PR so far! âœ¨ Unfortunately, the [Travis CI build](https://travis-ci.org/JasonEtco/public-test/builds/123?utm_source&#x3D;github_status&amp;utm_medium&#x3D;notification) is failing. Here's the output:

##### \`npm test\`

\`\`\`
> public-test@1.0.0 test /home/travis/build/JasonEtco/public-test
> echo \\"Error: no test specified\\" && exit 1

Error: no test specified
\`\`\`

I'm sure you can fix it! If you need help, don't hesitate to ask a maintainer of the project!

---

###### This comment was automagically generated by [ci-reporter](https://github.com/apps/ci-reporter). If you see a problem, open [an issue here](https://github.com/jasonetco/ci-reporter)."
`;
